File: cmd/create/tag.go
package create

import (
	"fmt"

	"github.com/fioncat/roxide/cmd"
	"github.com/fioncat/roxide/pkg/context"
	"github.com/fioncat/roxide/pkg/git"
	"github.com/fioncat/roxide/pkg/repoutils"
	"github.com/fioncat/roxide/pkg/term"
	"github.com/spf13/cobra"
)

func newTag() *cobra.Command {
	var opts tagOptions

	c := &cobra.Command{
		Use:   "tag [NAME]",
		Short: "Create a new tag",

		Args: cobra.MaximumNArgs(1),

		ValidArgsFunction: cmd.NoneCompletion,
	}

	c.Flags().StringVarP(&opts.rule, "rule", "r", "", "The rule to apply on the tag")

	return cmd.Build(c, &opts)
}

type tagOptions struct {
	name string

	rule string
}

func (o *tagOptions) Complete(c *cobra.Command, args []string) error {
	if len(args) > 0 {
		o.name = args[0]
	}
	return nil
}

func (o *tagOptions) Run(ctx *context.Context) error {
	repo, err := repoutils.MustGetCurrentRepo(ctx)
	if err != nil {
		return err
	}

	err = ctx.SetRepo(repo)
	if err != nil {
		return err
	}

	if o.name != "" {
		return o.createTag(ctx, o.name)
	}

	tag, err := git.GetLatestTag(ctx.GetRepoPath())
	if err != nil {
		return err
	}

	var rule string
	if o.rule != "" {
		var ok bool
		rule, ok = ctx.Config.GetTagRule(o.rule)
		if !ok {
			return fmt.Errorf("cannot find tag rule %q in config", o.rule)
		}
	} else {
		rules := ctx.Config.TagRules
		if len(rules) == 0 {
			return fmt.Errorf("no tag rule found in config")
		}

		items := make([]string, 0, len(rules))
		for _, rule := range rules {
			items = append(items, rule.Name)
		}

		idx, err := ctx.Selector.Select(items)
		if err != nil {
			return err
		}

		rule = rules[idx].Rule
	}

	newTag, err := tag.ApplyRule(rule)
	if err != nil {
		return err
	}

	term.PrintInfo("Apply rule %q on %q", rule, tag)
	err = term.Confirm("Do you want to create tag %q", newTag)
	if err != nil {
		return err
	}

	return o.createTag(ctx, string(newTag))
}

func (o *tagOptions) createTag(ctx *context.Context, tag string) error {
	gitCmd := git.WithPath(ctx.GetRepoPath())

	err := gitCmd.Run("tag", tag)
	if err != nil {
		return err
	}

	err = term.Confirm("Do you want to push tag %q to remote", tag)
	if err == nil {
		return gitCmd.Run("push", "origin", tag)
	}

	return nil
}

File: src/cmd/create/tag.rs
use anyhow::Result;
use async_trait::async_trait;
use clap::Args;
use console::style;

use crate::cmd::Command;
use crate::cmd::complete::{CompleteArg, CompleteCommand, funcs};
use crate::config::context::ConfigContext;
use crate::exec::git::tag::{Tag, UpdateTagRule};
use crate::{confirm, debug, info, output};

/// Create a new tag and push it to the remote.
#[derive(Debug, Args)]
pub struct CreateTagCommand {
    /// Two forms: tag name or tag rule. Available rules: [patch, minor, major, date,
    /// date-dash, date-dot].
    /// When using rules, applies the rule to the latest tag to generate a new tag.
    /// First three rules are for semantic versioning, last three for date-based tags.
    /// Examples: patch: v0.4.2 -> v0.4.3, minor: v1.5.6 -> v1.6.0, major: 2.3.4 -> 3.0.0,
    /// date-dash: 2024-01-02, date-dot: 2024.01.02, date: auto-detects dash or dot format
    /// from latest tag.
    /// If not one of these rules, creates the tag directly. For example: "v1.2.3", "1.4.0"
    pub tag: String,
}

#[async_trait]
impl Command for CreateTagCommand {
    fn name() -> &'static str {
        "tag"
    }

    async fn run(self, ctx: ConfigContext) -> Result<()> {
        debug!("[cmd] Run create tag command: {:?}", self);

        let tag = match UpdateTagRule::from_str(&self.tag) {
            Some(rule) => {
                let latest = Tag::get_latest(ctx.git())?;
                info!("Latest tag is: {}", style(&latest.name).green());
                let new_tag = rule.apply(&latest.name)?;
                confirm!("Do you want to create tag {}", style(&new_tag).magenta());
                new_tag
            }
            None => {
                let tags = Tag::list(ctx.git())?;
                for t in tags {
                    if t.name == self.tag {
                        output!("Tag {:?} already exists", self.tag);
                        return Ok(());
                    }
                }
                self.tag
            }
        };

        ctx.git()
            .execute(["tag", &tag], format!("Creating tag {tag:?}"))?;

        ctx.git().execute(
            ["push", "origin", &tag],
            format!("Pushing tag {tag:?} to remote"),
        )?;

        Ok(())
    }

    fn complete() -> CompleteCommand {
        Self::default_complete().arg(CompleteArg::new().complete(funcs::complete_tag_method))
    }
}
